sw/host/ot_certs/src/asn1/codegen.rs

```diff
@@ -0,0 +1,370 @@
+                let data = [vec![0; size - data.len()], data.clone()].concat();
+                let const_name = self.add_constant_byte_array(name_hint, &data);
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_bytes(&state, {const_name}, sizeof({const_name})));\n"));
+            }
+            Value::Variable(Variable { name, .. }) => {
+                // TODO: handle conv
+                // TODO: check size/pad
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_bytes(&state, values->{name}, values->{name}_size));\n"));
+            }
+        }
+        Ok(())
+    }
+
+    /// Push a byte array of fixed length into the ASN1 output. Note that this function does not add a tag to
+    /// the ASN1 output.
+    fn push_byte_array(&mut self, name_hint: Option<String>, val: &Value<Vec<u8>>) -> Result<()> {
+        match val {
+            Value::Literal(x) => {
+                let const_name = self.add_constant_byte_array(name_hint, x);
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_bytes(&state, {const_name}, sizeof({const_name})));\n"));
+            }
+            Value::Variable(Variable { name, .. }) => {
+                // TODO: handle conv
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_bytes(&state, values->{name}, values->{name}_size));\n"));
+            }
+        }
+        Ok(())
+    }
+
+    /// Push an optionally tagged string into the ASN1 output.
+    fn push_string(
+        &mut self,
+        _name_hint: Option<String>,
+        str_type: &Tag,
+        val: &Value<String>,
+    ) -> Result<()> {
+        let str_type = str_type.codestring();
+        match val {
+            Value::Literal(x) => {
+                // NOTE: we are not using a constant for the size of the string because technically there is no way
+                // to know what the `char` type is and therefore what is the length of this string for the C compiler.
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_string(&state, {str_type}, \"{x}\", strlen(\"{x}\")));\n"));
+            }
+            Value::Variable(Variable { name, .. }) => {
+                // TODO: handle conv
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_string(&state, {str_type}, values->{name}, values->{name}_len));\n"));
+            }
+        }
+        Ok(())
+    }
+
+    fn push_oid(&mut self, oid: &Oid) -> Result<()> {
+        // Create constant.
+        let oid_const_name =
+            self.add_constant_byte_array(Some(format!("oid_{}", oid)), &oid.to_der()?);
+        self.push_raw_oid(&oid_const_name, &format!("sizeof({oid_const_name})"));
+
+        Ok(())
+    }
+
+    // Helper functions for outputting ASN1 tags.
+
+    fn push_tag(
+        &mut self,
+        name_hint: Option<String>,
+        tag: &Tag,
+        gen: &dyn Fn(&mut dyn Builder) -> Result<()>,
```

Comment:
Do we need dynamic dispatch for `Builder`? It seems that `&mut Self` is enough?

Reply:
I have decided to keep the dynamic dispatch for now. I plan to have another implementation that directly outputs bytes so that we can compare the asn1 library with a rust reference for unittest. This can be changed in the future if we want.